plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -238,10 +250,33 @@ public ConnectorPageSource createPageSource(
         Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(partitionData, partitionSpec);
 
         ImmutableList.Builder<IcebergColumnHandle> requiredColumnsBuilder = ImmutableList.builder();
+        Set<Integer> requiredColumnIds = new HashSet<>();
         requiredColumnsBuilder.addAll(icebergColumns);
+        icebergColumns.stream().map(IcebergColumnHandle::getId).forEach(requiredColumnIds::add);
         deleteFilterRequiredSchema.stream()
                 .filter(column -> !icebergColumns.contains(column))
-                .forEach(requiredColumnsBuilder::add);
+                .forEach(column -> {
+                    requiredColumnsBuilder.add(column);
+                    requiredColumnIds.add(column.getId());
+                });
+        icebergColumns.stream().filter(IcebergColumnHandle::isUpdateRowIdColumn).findFirst().ifPresent(updateRowIdColumn -> {
+            updateRowIdColumn.getColumnIdentity().getChildren().forEach(requiredColumnIdentity -> {
+                // The $row_id column is not read directly, but its children are
```

Comment:
Something is off about this sentence. Please fix the grammar so that it makes sense.